File: backend/app/schemas/batch_import.py
"""
Batch Import schemas for API requests and responses
"""

import re
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator


class BatchImportUploadResponse(BaseModel):
    """Response after uploading batch import file"""

    batch_id: int = Field(..., description="批次匯入ID")
    file_name: str = Field(..., description="檔案名稱")
    total_records: int = Field(..., description="總筆數")
    preview_data: List[Dict[str, Any]] = Field(..., description="預覽資料")
    validation_summary: Dict[str, Any] = Field(..., description="驗證摘要")


class BatchImportValidationError(BaseModel):
    """Single validation error"""

    row_number: int = Field(..., description="行號")
    student_id: Optional[str] = Field(None, description="學號")
    field: str = Field(..., description="欄位名稱")
    error_type: str = Field(..., description="錯誤類型")
    message: str = Field(..., description="錯誤訊息")


class BatchImportConfirmRequest(BaseModel):
    """Request to confirm batch import"""

    batch_id: int = Field(..., description="批次匯入ID")
    confirm: bool = Field(..., description="確認匯入")


class BatchImportConfirmResponse(BaseModel):
    """Response after confirming batch import"""

    batch_id: int
    success_count: int = Field(..., description="成功筆數")
    failed_count: int = Field(..., description="失敗筆數")
    errors: List[BatchImportValidationError] = Field(default=[], description="錯誤清單")
    created_application_ids: List[int] = Field(default=[], description="成功建立的申請ID列表")


class BatchImportHistoryItem(BaseModel):
    """Single batch import history item"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    college_code: str
    scholarship_type_id: Optional[int] = None
    academic_year: int
    semester: Optional[str] = None
    file_name: str
    total_records: int
    success_count: int
    failed_count: int
    import_status: str
    created_at: datetime
    importer_name: Optional[str] = None  # From relationship


class BatchImportHistoryResponse(BaseModel):
    """Batch import history response"""

    total: int = Field(..., description="總筆數")
    items: List[BatchImportHistoryItem] = Field(..., description="匯入歷史列表")


class BatchImportDetailResponse(BaseModel):
    """Detailed batch import information"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    college_code: str
    scholarship_type_id: Optional[int] = None
    academic_year: int
    semester: Optional[str] = None
    file_name: str
    file_path: Optional[str] = None
    total_records: int
    success_count: int
    failed_count: int
    error_summary: Optional[Dict[str, Any]] = None
    import_status: str
    created_at: datetime
    updated_at: Optional[datetime] = None
    importer_name: Optional[str] = None
    created_applications: Optional[List[int]] = Field(default=[], description="建立的申請ID列表")


class ApplicationDataRow(BaseModel):
    """Single row of application data from Excel/CSV"""

    # 必要欄位
    student_id: str = Field(..., description="學號", min_length=1, max_length=20)
    student_name: str = Field(..., description="姓名", min_length=1, max_length=100)

    # 子類型
    sub_types: List[str] = Field(default=[], description="子類型列表")

    # 郵局帳戶資訊
    postal_account: Optional[str] = Field(None, description="郵局帳號", max_length=20)

    # 續領
    is_renewal: bool = Field(default=False, description="是否為續領申請")
    renewal_year: Optional[int] = Field(default=None, description="續領年份，例如 113")

    # 其他動態欄位
    custom_fields: Optional[Dict[str, Any]] = Field(default={}, description="其他自定義欄位")

    @field_validator("student_id")
    @classmethod
    def validate_student_id(cls, v: str) -> str:
        """Validate student ID format - alphanumeric only"""
        v = v.strip()
        if not re.match(r"^[A-Za-z0-9]+$", v):
            raise ValueError("學號僅能包含英文字母和數字")
        return v

    @field_validator("student_name")
    @classmethod
    def validate_name_fields(cls, v: str) -> str:
        """Validate name fields - no HTML/script tags for XSS prevention"""
        v = v.strip()
        # Check for common XSS patterns
        if re.search(r"<[^>]*script|<[^>]*iframe|javascript:", v, re.IGNORECASE):
            raise ValueError("名稱欄位包含不允許的字元")
        return v

    @field_validator("postal_account")
    @classmethod
    def validate_postal_account(cls, v: Optional[str]) -> Optional[str]:
        """Validate postal account - numbers and hyphens only"""
        if v is None:
            return v
        v = v.strip()
        if not re.match(r"^[0-9\-]+$", v):
            raise ValueError("郵局帳號僅能包含數字和連字號")
        return v


class BatchImportDataRequest(BaseModel):
    """Request to create batch import with data"""

    scholarship_type: str = Field(..., description="獎學金類型代碼")
    academic_year: int = Field(..., description="學年度", ge=100, le=200)
    semester: Optional[str] = Field(None, description="學期")
    data_rows: List[ApplicationDataRow] = Field(..., description="申請資料列表", min_length=1)


class BatchImportUpdateRecordRequest(BaseModel):
    """Request to update a single record in batch import"""

    record_index: int = Field(..., description="記錄索引（從0開始）", ge=0)
    updates: Dict[str, Any] = Field(..., description="要更新的欄位")


class BatchImportRevalidateResponse(BaseModel):
    """Response after revalidation"""

    batch_id: int
    total_records: int
    valid_count: int
    invalid_count: int
    errors: List[Dict[str, Any]] = Field(default=[], description="驗證錯誤列表")
    warnings: List[Dict[str, Any]] = Field(default=[], description="驗證警告列表")


class BatchDocumentUploadResult(BaseModel):
    """Result of document matching"""

    student_id: str
    file_name: str
    document_type: str
    status: str  # "success" or "error"
    message: Optional[str] = None
    application_id: Optional[int] = None


class BatchDocumentUploadResponse(BaseModel):
    """Response after bulk document upload"""

    batch_id: int
    total_files: int
    matched_count: int
    unmatched_count: int
    error_count: int
    results: List[BatchDocumentUploadResult] = Field(default=[], description="上傳結果詳情")

File: frontend/lib/api/modules/batch-import.ts
/**
 * Batch Import API Module (OpenAPI-typed)
 *
 * Handles bulk application data import for college administrators:
 * - Upload and validate Excel/CSV files
 * - Preview data before confirmation
 * - Confirm batch imports
 * - View import history and details
 * - Download import templates
 *
 * Now using openapi-fetch for full type safety from backend OpenAPI schema
 */

import { typedClient } from '../typed-client';
import { toApiResponse } from '../compat';
import { createFileUploadFormData, type MultipartFormData } from '../form-data-helpers';
import type { ApiResponse } from '../../api.legacy';

type BatchUploadResult = {
  batch_id: number;
  file_name: string;
  total_records: number;
  preview_data: Array<Record<string, any>>;
  validation_summary: {
    valid_count: number;
    invalid_count: number;
    warnings: string[];
    errors: Array<{
      row: number;
      field?: string;
      message: string;
    }>;
  };
};

type BatchConfirmResult = {
  success_count: number;
  failed_count: number;
  errors: Array<{
    row: number;
    student_id: string;
    error: string;
  }>;
  created_application_ids: number[];
};

type BatchUpdateRecordResult = {
  updated_record: Record<string, any>;
};

type BatchRevalidateResult = {
  batch_id: number;
  total_records: number;
  valid_count: number;
  invalid_count: number;
  errors: Array<{
    row: number;
    field: string;
    message: string;
  }>;
};

type BatchDeleteRecordResult = {
  deleted_record: Record<string, any>;
  remaining_records: number;
};

type BatchDocumentUploadResult = {
  student_id: string;
  file_name: string;
  document_type: string;
  status: string;
  message?: string;
  application_id?: number;
};

type BatchDocumentUploadResponse = {
  batch_id: number;
  total_files: number;
  matched_count: number;
  unmatched_count: number;
  error_count: number;
  results: BatchDocumentUploadResult[];
};

type BatchHistoryItem = {
  id: number;
  file_name: string;
  importer_name?: string;
  created_at: string;
  total_records: number;
  success_count: number;
  failed_count: number;
  import_status: string;
  scholarship_type_id?: number;
  college_code: string;
  academic_year: number;
  semester: string | null;
};

type BatchHistoryResponse = {
  items: BatchHistoryItem[];
  total: number;
};

type BatchDetails = BatchHistoryItem & {
  created_applications?: number[];
  validation_summary: {
    valid_count: number;
    invalid_count: number;
    warnings: string[];
    errors: Array<{
      row: number;
      field?: string;
      message: string;
    }>;
  };
  preview_data: Array<Record<string, any>>;
  processing_errors: Array<{
    row: number;
    student_id: string;
    error: string;
  }>;
};

export function createBatchImportApi() {
  return {
    /**
     * Upload and validate batch import data file
     * Type-safe: FormData upload with query parameters validated
     */
    uploadData: async (
      file: File,
      scholarshipType: string,
      academicYear: number,
      semester: string
    ): Promise<ApiResponse<BatchUploadResult>> => {
      const formData = createFileUploadFormData({ file });

      const response = await typedClient.raw.POST('/api/v1/college-review/batch-import/upload-data', {
        params: {
          query: {
            scholarship_type: scholarshipType,
            academic_year: academicYear,
            semester: semester,
          },
        },
        body: formData as MultipartFormData<{ file: string }>,
      });
      return toApiResponse<BatchUploadResult>(response);
    },

    /**
     * Confirm batch import after validation
     * Type-safe: Path parameter and request body validated against OpenAPI
     */
    confirm: async (
      batchId: number,
      confirm: boolean = true
    ): Promise<ApiResponse<BatchConfirmResult>> => {
      const response = await typedClient.raw.POST('/api/v1/college-review/batch-import/{batch_id}/confirm', {
        params: { path: { batch_id: batchId } },
        body: { batch_id: batchId, confirm },
      });
      return toApiResponse<BatchConfirmResult>(response);
    },

    /**
     * Update a single record in the batch import
     * Type-safe: Path parameter and request body validated against OpenAPI
     */
    updateRecord: async (
      batchId: number,
      recordIndex: number,
      updates: Record<string, any>
    ): Promise<ApiResponse<BatchUpdateRecordResult>> => {
      const response = await typedClient.raw.PATCH('/api/v1/college-review/batch-import/{batch_id}/records', {
        params: { path: { batch_id: batchId } },
        body: { record_index: recordIndex, updates } as any,
      });
      return toApiResponse<BatchUpdateRecordResult>(response);
    },

    /**
     * Re-validate all records in the batch import
     * Type-safe: Path parameter validated against OpenAPI
     */
    revalidate: async (
      batchId: number
    ): Promise<ApiResponse<BatchRevalidateResult>> => {
      const response = await typedClient.raw.POST('/api/v1/college-review/batch-import/{batch_id}/validate', {
        params: { path: { batch_id: batchId } },
      });
      return toApiResponse<BatchRevalidateResult>(response);
    },

    /**
     * Delete a single record from the batch import
     * Type-safe: Path parameters validated against OpenAPI
     */
    deleteRecord: async (
      batchId: number,
      recordIndex: number
    ): Promise<ApiResponse<BatchDeleteRecordResult>> => {
      const response = await typedClient.raw.DELETE('/api/v1/college-review/batch-import/{batch_id}/records/{record_index}', {
        params: { path: { batch_id: batchId, record_index: recordIndex } },
      });
      return toApiResponse<BatchDeleteRecordResult>(response);
    },

    /**
     * Upload documents in bulk for batch import (ZIP file)
     * Type-safe: Path parameter and FormData body properly typed
     */
    uploadDocuments: async (
      batchId: number,
      zipFile: File
    ): Promise<ApiResponse<BatchDocumentUploadResponse>> => {
      const formData = createFileUploadFormData({ file: zipFile });

      const response = await typedClient.raw.POST('/api/v1/college-review/batch-import/{batch_id}/documents', {
        params: { path: { batch_id: batchId } },
        body: formData as MultipartFormData<{ file: string }>,
      });
      return toApiResponse<BatchDocumentUploadResponse>(response);
    },

    /**
     * Get batch import history with optional filtering
     * Type-safe: Query parameters validated against OpenAPI
     */
    getHistory: async (params?: {
      skip?: number;
      limit?: number;
      status?: string;
    }): Promise<ApiResponse<BatchHistoryResponse>> => {
      const response = await typedClient.raw.GET('/api/v1/college-review/batch-import/history', {
        params: {
          query: {
            skip: params?.skip,
            limit: params?.limit,
            status: params?.status,
          },
        },
      });
      return toApiResponse<BatchHistoryResponse>(response);
    },

    /**
     * Get detailed information about a specific batch import
     * Type-safe: Path parameter validated against OpenAPI
     */
    getDetails: async (
      batchId: number
    ): Promise<ApiResponse<BatchDetails>> => {
      const response = await typedClient.raw.GET('/api/v1/college-review/batch-import/{batch_id}/details', {
        params: { path: { batch_id: batchId } },
      });
      return toApiResponse<BatchDetails>(response);
    },

    /**
     * Download batch import template for a scholarship type
     * Type-safe: Returns blob for file download
     */
    downloadTemplate: async (scholarshipType: string): Promise<void> => {
      const token = typedClient.getToken();
      const baseURL = typeof window !== "undefined" ? "" : process.env.INTERNAL_API_URL || "http://localhost:8000";

      const response = await fetch(
        `${baseURL}/api/v1/college-review/batch-import/template?scholarship_type=${encodeURIComponent(scholarshipType)}`,
        {
          method: "GET",
          headers: {
            ...(token && { Authorization: `Bearer ${token}` }),
          },
        }
      );

      if (!response.ok) {
        throw new Error("Failed to download template");
      }

      const blob = await response.blob();
      const url = window.URL.createObjectURL(blob);
      const link = document.createElement("a");
      link.href = url;

      // Extract filename from Content-Disposition header
      const contentDisposition = response.headers.get("content-disposition");
      let filename = `batch_import_template_${scholarshipType}.xlsx`;
      if (contentDisposition) {
        // Match filename*=UTF-8''encoded_name (RFC 5987)
        const filenameMatch = contentDisposition.match(/filename\*=UTF-8''([^;]+)/);
        if (filenameMatch) {
          filename = decodeURIComponent(filenameMatch[1].trim());
        }
      }

      link.download = filename;
      document.body.appendChild(link);
      link.click();
      document.body.removeChild(link);
      window.URL.revokeObjectURL(url);
    },

    /**
     * Delete a batch import and all its related applications
     * Type-safe: Path parameter validated against OpenAPI
     */
    deleteBatch: async (batchId: number): Promise<ApiResponse<{ batch_id: number; deleted_applications: number }>> => {
      const response = await typedClient.raw.DELETE('/api/v1/college-review/batch-import/{batch_id}', {
        params: { path: { batch_id: batchId } },
      });
      return toApiResponse<{ batch_id: number; deleted_applications: number }>(response);
    },

    /**
     * Download original file for a batch import
     * Type-safe: Returns blob for file download
     */
    downloadFile: async (batchId: number): Promise<void> => {
      const token = typedClient.getToken();
      const baseURL =
        typeof window !== "undefined"
          ? ""
          : process.env.INTERNAL_API_URL || "http://localhost:8000";

      const response = await fetch(
        `${baseURL}/api/v1/college-review/batch-import/${batchId}/download`,
        {
          method: "GET",
          headers: {
            ...(token && { Authorization: `Bearer ${token}` }),
          },
        }
      );

      if (!response.ok) {
        throw new Error("Failed to download file");
      }

      const blob = await response.blob();
      const url = window.URL.createObjectURL(blob);
      const link = document.createElement("a");
      link.href = url;

      // Extract filename from Content-Disposition header
      const contentDisposition = response.headers.get("content-disposition");
      let filename = `batch_import_${batchId}.xlsx`;
      if (contentDisposition) {
        // Match filename*=UTF-8''encoded_name (RFC 5987)
        const filenameMatch = contentDisposition.match(
          /filename\*=UTF-8''([^;]+)/
        );
        if (filenameMatch) {
          filename = decodeURIComponent(filenameMatch[1].trim());
        }
      }

      link.download = filename;
      document.body.appendChild(link);
      link.click();
      document.body.removeChild(link);
      window.URL.revokeObjectURL(url);
    },
  };
}
